index.js

```diff
@@ -124,19 +124,31 @@ function Encoder() {}
  */
 
 Encoder.prototype.encode = function(obj, callback){
-  if ((obj.type === exports.EVENT || obj.type === exports.ACK) && hasBin(obj.data)) {
-    obj.type = obj.type === exports.EVENT ? exports.BINARY_EVENT : exports.BINARY_ACK;
+
+  /**
+   * For object data with circular references, hasBin will throw
+   * exceptions, which must be handled or it could hang the channel.
+   * And the circular references only happen to non-Binary data, 
+   * so we will invoke callback with an array.
+   */
+  try {
+    var hasBinary = hasBin(obj.data);
```

Comment:
Currently, `hasBin` is only called when the first condition `(obj.type === exports.EVENT || obj.type === exports.ACK)` is true, but it will always be called here right?

Reply:
Good point, I think the try-catch block should be moved into the if statements that checks `obj.type`. I will update this code tonight.